### ISSUE TITLE
levenshtein distance and jaro winkler returning inverted value

### ISSUE THREAD

**Issue by funkysandman** [moj-analytical-services/splink#951]
### What happens?

In my databricks environment I've been experiencing the opposite expected values when doing jaro_winkler comparisons

Probably an issue on my end, but it really messes up training as you could imagine

### To Reproduce

in spark, doing a simple select in a notebook like this:

from pyspark.sql import types
from pyspark.sql.types import  DoubleType
spark.udf.registerJavaFunction('jaro_winkler', 'uk.gov.moj.dash.linkage.JaroWinklerSimilarity', types.DoubleType())

%sql
select jaro_winkler("Splink program", "Splink produce") 

 
jaro_winkler(Splink program, Splink produce)

**0.11428571428571421**
Showing 1 row.

0.18 seconds runtime

### OS:

11.3 LTS (includes Apache Spark 3.3.0, Scala 2.12)

### Splink version:

3.4.4

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by mamonu** [comment]
That is strange.
I loaded the jar (version 0.0.9) and then run the same code and it looks like im getting the correct similarity.

![image](https://user-images.githubusercontent.com/3820011/208213766-1e9d115a-8a1c-4e3f-a3f8-f3269c3d183b.png)


Can you please check what jar you are importing and using (it should be available as configuration in spark.jars
by running the following code

```python
configurations = spark.sparkContext.getConf().getAll()
for key,item in configurations: 
    if key == "spark.jars":
        print (item)
```

The path should be different than ours but this will show the version of the jar thats being used.
Hopefully it's 0.0.9. Then we can take it from there.


Also you are saying this happens with levenstein distance?

**Comment by funkysandman** [comment]
![image](https://user-images.githubusercontent.com/37027650/208214580-d9ac92b4-1bb4-476b-8763-0927ddc88d88.png)

Actually I didn't use the levenstein distance, just the jaro_winkler_sim

**Comment by mamonu** [comment]
can you try the same by using the 0.0.9 version taken from here?

https://github.com/moj-analytical-services/splink_scalaudfs/tree/main/jars

**Comment by funkysandman** [comment]
same result (no change) with 0.0.9 and splink 3.5.1

**Comment by zzandi** [comment]
Hi, we have experienced the same issue. This is due to spark version incompatibility, and it happens in spark 3.3. However, since spark 3.2 has considerably slower performance, instead of downgrading one could rebuild the jar file with a new version of “org.apache.commons” 1.10.0 instead of 1.4 and also add org.apache.spark 3.3 dependency to pom.xml to overcome this problem.

**Comment by mamonu** [comment]
@zzandi  this is very good to know. I will update the jar at some point and will definitely take all this into account.
Perhaps offere 2 sets of jars one for older sparks (2.4.x etc) and one for the latest spark (3.3.x)

**Comment by mshearer0** [comment]
I get the same result:

```
+--------------------------------------------+
|jaro_winkler(Splink Program, Splink Produce)|
+--------------------------------------------+
|0.11428571428571421                         |
+--------------------------------------------+
```
Using Spark 3.1.3 and scala-udf-similarity-0.0.9.jar on Google Cloud Dataproc.

**Comment by mshearer0** [comment]
> > with a new version of “org.apache.commons” 1.10.0 instead of 1.4 and also add org.apache.spark 3.3 dependency to pom.x
> 
> ok, I suspected something was up with the underlying libraries.. in the meantime I adjusted my calculations to be 1-jaro_sim and reversed my comparison logic in the settings

@mamonu - Could you please share how you did this. I tried to use a custom sql condition but couldn't find a _sim function.

**Comment by mamonu** [comment]
ok so updated the versions on pom.xml version and re-built the jar


https://github.com/moj-analytical-services/splink_scalaudfs/blob/spark3_x/jars/scala-udf-similarity-0.1.0_spark3.3.jar

@mshearer0 @funkysandman  or anyone that hasnt used a recompiled verison can you try now with this jar and let me know of your results?

 Also what people have done is create a    1-jaro_winkler( col1,col2) expression to deal with this. you can name it as you wish when you register the function.

@zzandi was in a rush hence he wrote in shorthand.


anyway.hopefully if people can confirm that this works on `google cloud` and `databricks` syste,s this 1-jw trick will not be needed . But i would love if you or other people with this issue tested this the new jar on your platforms  and reported back

an example where this jar works is here:

https://gist.github.com/mamonu/4a133a98abfc10c9797bd83045b464b0

**Comment by mshearer0** [comment]
Thanks @mamonu

On Spark 3.1.3 on Google Cloud dataproc i still get the same inverted result using the new jar:

+----------------------------+
|jaro_winkler(MARHTA, MARTHA)|
+----------------------------+
|0.03888888888888886         |
+----------------------------+

Is this the same issue as https://github.com/moj-analytical-services/splink_scalaudfs/issues/3 ?

**Comment by funkysandman** [comment]
first try at this in Azure databricks: 12.0 ML (includes Apache Spark 3.3.1, Scala 2.12)

jaro_winkler(RIJSWIJK, RIJSWIJK ZH)
1
0.05454545454545445

**Comment by zzandi** [comment]
I just noticed that you are using JaroWinklerDistance in your Scala code, could you please replace it with JaroWinklerSimlarity.
The following is working for me.  Databricks 11.3 LTS (includes Apache Spark 3.3.0, Scala 2.12),  commons text version 1.10,

class JaroWinklerSimilarity extends UDF2[String, String, Double] {
  override def call(left: String, right: String): Double = {
    // This has to be instantiated here (i.e. on the worker node)

    if ((left != null) & (right != null)) {
      val distance = new similarity.JaroWinklerSimilarity()
      distance(left, right)
    } else {
      0.0
    }
  }
}

object JaroWinklerSimilarity {
  def apply(): JaroWinklerSimilarity = {
    new JaroWinklerSimilarity()
  }
}

![image](https://user-images.githubusercontent.com/114579526/211766941-0d5962d9-5251-4e65-9a5a-6ce9e10d2ce6.png)

**Comment by mamonu** [comment]
```
• up to version 1.10, JaroWinklerDistance had a bug and it returned the same value as JaroWinklerSimilarity. (Please see
https://issues.apache.org/jira/browse/TEXT-191
```

I knew that this bug existed hence thats why before 1.10 it didnt matter . But yes this makes things much clearer now.

Thanks for doing this investigation about https://issues.apache.org/jira/browse/TEXT-191 . After 4 years of watching this i just didnt think it would be fixed! 

will be creating a new jar based on this information.

**Comment by mamonu** [comment]
@funkysandman @zzandi @mshearer0 and anyone else that seemed to have an issue.
I updated

https://github.com/moj-analytical-services/splink_scalaudfs/blob/spark3_x/jars/scala-udf-similarity-0.1.0_spark3.3.jar

can you please try again to test it???.
Thanks again for the testing as its the only way to know if something done works on platforms different to ours. 

Once I know this works I can actually merge it to main etc. (and never speak of it again as its quite embarrassing 😅 )

**Comment by funkysandman** [comment]
@mamonu 
looks like it's fixed for databricks in Azure.
(12.0 ML (includes Apache Spark 3.3.1, Scala 2.12))
 
jaro_winkler(cat, cats)
1
0.9416666666666667

thank you!

**Comment by mshearer0** [comment]
Yes, fixed for me too!, on Spark 3.1.3

Stage 1:>                                                          (0 + 1) / 1]
+----------------------------+
|jaro_winkler(MARHTA, MARTHA)|
+----------------------------+
|0.9611111111111111          |
+----------------------------+
